### PR TITLE
[test] Fix canary failures for gpu images

### PR DESCRIPTION
**Pull request by shantanutrip**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description
GPU images in canary do not have the cuda version in their tags. As a result, the function `def get_dockerfile_path_for_image(image_uri):` fails since it is not able to fetch the correct cuda version. This PR ensures that the failure does not happen for such GPU images. It uses the CUDA_VERSION environment variable in the image to extract the exact cuda version.

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/test_utils/__init__.py [hunk 1]
```diff
@@ -1210,6 +1210,17 @@ def get_cuda_version_from_tag(image_uri):
     if all(keyword in image_uri for keyword in cuda_str):
         cuda_framework_version = re.search(r"(cu\d+)-", image_uri).groups()[0]
 
+    if not cuda_framework_version and "gpu" in image_uri:
+        response_output = json.loads(run(f"docker inspect {image_uri}", hide=True).stdout)
+        env_vars = response_output[0]['ContainerConfig']['Env']
+        cuda_version_env_variables = [env_var for env_var in env_vars if env_var.startswith("CUDA_VERSION=")]
+        if not cuda_version_env_variables:
+            raise RuntimeError(f"Image {image_uri} does not have an Env variable called CUDA_VERSION")
```
Comment (saimidu) [inline test/test_utils/__init__.py:1218]: Why is this check being conducted in a `get` function for the CUDA version being obtained from an image tag?

This function is not a test, and the lack of `CUDA_VERSION` as an env variable in the image should not raise a `RuntimeError` here.

File: test/test_utils/__init__.py [hunk 2]
```diff
@@ -1210,6 +1210,17 @@ def get_cuda_version_from_tag(image_uri):
     if all(keyword in image_uri for keyword in cuda_str):
         cuda_framework_version = re.search(r"(cu\d+)-", image_uri).groups()[0]
 
+    if not cuda_framework_version and "gpu" in image_uri:
```
Comment (saimidu) [inline test/test_utils/__init__.py:1213]: Why not just get the CUDA version from the other tags that **will** be on the image?

Do we know for sure that `CUDA_VERSION` is one of the env variables created in GPU images?

File: test/dlc_tests/sanity/test_security_check.py [hunk 3]
```diff
@@ -101,11 +101,20 @@ def test_ecr_scan(image, ecr_client, sts_client, region):
     """
     test_account_id = sts_client.get_caller_identity().get("Account")
     image_account_id = get_account_id_from_image_uri(image)
+    image_repo_uri, image_tag = image.split(":")
+    _, image_repo_name = image_repo_uri.split("/")
```
Comment (saimidu) [inline test/dlc_tests/sanity/test_security_check.py:105]: Please use existing utility functions such as `get_repository_and_tag_from_image_uri` to obtain the `image_repo_name` and `image_tag`, or create a new utility function to get the `image_repo_uri` from an ECR URI, rather than string-parsing right here.

File: test/test_utils/__init__.py [hunk 4]
```diff
@@ -1205,21 +1205,14 @@ def get_cuda_version_from_tag(image_uri):
...
-        cuda_version = cuda_version_string.split("=")[1]
-        image_cuda_version = ''.join(cuda_version.split('.')[:2])
-        cuda_framework_version = f"cu{image_cuda_version}"
+    response_output = json.loads(run(f"docker inspect {image_uri}", hide=True).stdout)
+    repo_tags = response_output[0]["RepoTags"]
+
+    for repo_tag in repo_tags:
+        if all(keyword in repo_tag for keyword in cuda_str):
+            cuda_framework_version = re.search(r"(cu\d+)-", repo_tag).groups()[0]
+            return cuda_framework_version
```
Comment (saimidu) [inline test/test_utils/__init__.py:1215]: 1. Please use `ecr.describe_images` rather than assuming that the image is downloaded.
2. Please add a `raise CudaVersionTagNotFoundException` or something if the for loop completes for a GPU image, and return `None` if the image is not a GPU image type.

File: test/test_utils/ecr.py [hunk 5]
```diff
@@ -268,3 +269,26 @@ def populate_ecr_scan_with_web_scraper_results(
...
+    """
+    account_id = image_uri.split('.')[0]
+    image_repo_uri, image_tag = image_uri.split(":")
+    _, image_repo_name = image_repo_uri.split("/")
+    response = ecr_client.describe_images(
+        registryId=account_id,
+        repositoryName=image_repo_name,
+        imageIds=[
+            {
+                'imageDigest': get_image_digest(image_uri),
```
Comment (saimidu) [inline test/test_utils/ecr.py:289]: 1. The `imageDigest` input is unnecessary if `imageTag` is supplied as a part of the `describe_images` call.
2. The `get_image_digest` function makes the assumption that the image has been pulled, when that doesn't necessarily have to be the case. We should not require a time-expensive operation such as pulling an entire image just to query the tags on the image on ECR.

File: test/test_utils/__init__.py [hunk 6]
```diff
@@ -1198,19 +1205,50 @@ def get_framework_from_image_uri(image_uri):
...
+    """
+    Uses ecr describe to generate all the tags of an image.
+    
+    :param ecr_client: boto3 Client for ECR
+    :param image_uri: str Image URI
+    :return: list, All the image tags
+    """
+    account_id = image_uri.split('.')[0]
+    image_repo_uri, image_tag = image_uri.split(":")
+    _, image_repo_name = image_repo_uri.split("/")
```
Comment (saimidu) [inline test/test_utils/__init__.py:1218]: Please use existing utility functions to get
account ID: `get_account_id_from_image_uri`
image repo and tag: `get_repository_and_tag_from_image_uri`

**Review by saimidu — DISMISSED** [review]
Approved. Please revert temp changes and merge.

**Review by saimidu — APPROVED** [review]
